File: vmlib/src/main/java/me/shouheng/vmlib/base/BaseViewModel.java
package me.shouheng.vmlib.base;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import java.util.List;

import me.shouheng.vmlib.anno.ViewModelConfiguration;
import me.shouheng.vmlib.bean.Resources;
import me.shouheng.vmlib.bus.Bus;
import me.shouheng.vmlib.holder.LiveDataHolder;

/**
 * Basic implementation of common ViewModel.
 *
 * The view model has no pre-defined model associated, for MVVMs don't want to take care of the
 * model logic. You can get the data source anywhere, which is perhaps useful for small project.
 *
 * @author <a href="mailto:shouheng2015@gmail.com">WngShhng</a>
 * @version 2019-6-29
 */
public class BaseViewModel extends AndroidViewModel {

    private LiveDataHolder holder = new LiveDataHolder();

    private LiveDataHolder listHolder = new LiveDataHolder<List>();

    /**
     * If the view model use event bus
     */
    private boolean useEventBus = false;

    public BaseViewModel(@NonNull Application application) {
        super(application);
    }

    {
        ViewModelConfiguration configuration = this.getClass().getAnnotation(ViewModelConfiguration.class);
        if (configuration != null) {
            useEventBus = configuration.useEventBus();
            if (useEventBus) {
                Bus.get().register(this);
            }
        }
    }

    /**
     * Get the LiveData of given type.
     *
     * @param dataType the data type.
     * @param <T>      the data type.
     * @return         the live data.
     */
    public <T> MutableLiveData<Resources<T>> getObservable(Class<T> dataType) {
        return holder.getLiveData(dataType, false);
    }

    /**
     * Get the LiveData of given type.
     *
     * @param dataType the data type.
     * @param single   is single event
     * @param <T>      the data type.
     * @return         the live data.
     * @see SingleLiveEvent
     */
    public <T> MutableLiveData<Resources<T>> getObservable(Class<T> dataType, boolean single) {
        return holder.getLiveData(dataType, single);
    }

    /**
     * Get the LiveData of given type and flag
     *
     * @param dataType the data type
     * @param flag     the flag
     * @param <T>      the generic data type.
     * @return         the live data.
     */
    public <T> MutableLiveData<Resources<T>> getObservable(Class<T> dataType, int flag) {
        return holder.getLiveData(dataType, flag, false);
    }

    /**
     * et the LiveData of given type and flag
     *
     * @param dataType the data type
     * @param flag     the flag
     * @param single   is single event
     * @param <T>      the generic data type.
     * @return         the live data.
     * @see SingleLiveEvent
     */
    public <T> MutableLiveData<Resources<T>> getObservable(Class<T> dataType, int flag, boolean single) {
        return holder.getLiveData(dataType, flag, single);
    }

    /**
     * Get the LiveData of given list type.
     *
     * @param dataType the data type of element in list.
     * @param <T>      the generic data type.
     * @return         the live data.
     */
    public <T> MutableLiveData<Resources<List<T>>> getListObservable(Class<T> dataType) {
        return listHolder.getLiveData(dataType, false);
    }

    /**
     * Get the LiveData of given list type.
     *
     * @param dataType the data type of element in list.
     * @param single   is single event
     * @param <T>      the generic data type.
     * @return         the live data.
     * @see SingleLiveEvent
     */
    public <T> MutableLiveData<Resources<List<T>>> getListObservable(Class<T> dataType, boolean single) {
        return listHolder.getLiveData(dataType, single);
    }

    /**
     * Get the LiveData of given list type and flag
     *
     * @param dataType the data type
     * @param flag     the flag
     * @param <T>      the generic data type.
     * @return         the live data.
     */
    public <T> MutableLiveData<Resources<List<T>>> getListObservable(Class<T> dataType, int flag) {
        return listHolder.getLiveData(dataType, flag, false);
    }

    /**
     * Get the LiveData of given list type and flag
     *
     * @param dataType the data type
     * @param flag     the flag
     * @param single   is single event
     * @param <T>      the generic data type.
     * @return         the live data.
     * @see SingleLiveEvent
     */
    public <T> MutableLiveData<Resources<List<T>>> getListObservable(Class<T> dataType, int flag, boolean single) {
        return listHolder.getLiveData(dataType, flag, single);
    }

    /**
     * Notify the observer the success state.
     *
     * @param dataType data type
     * @param data     data
     * @param <T>      data type
     */
    public <T> void setSuccess(Class<T> dataType, T data) {
        getObservable(dataType).setValue(Resources.success(data));
    }

    /** @see #setSuccess(Class, Object) */
    public <T> void setLoading(Class<T> dataType) {
        getObservable(dataType).setValue(Resources.<T>loading());
    }

    /** @see #setSuccess(Class, Object) */
    public <T> void setFailed(Class<T> dataType, String code, String message) {
        getObservable(dataType).setValue(Resources.<T>failed(code, message));
    }

    /**
     * Notify the observer the success state.
     *
     * @param dataType data type
     * @param data     data
     * @param <T>      data type
     */
    public <T> void setListSuccess(Class<T> dataType, List<T> data) {
        getListObservable(dataType).setValue(Resources.success(data));
    }

    /** @see #setListSuccess(Class, List) */
    public <T> void setListLoading(Class<T> dataType) {
        getListObservable(dataType).setValue(Resources.<List<T>>loading());
    }

    /** @see #setListSuccess(Class, List) */
    public <T> void setListFailed(Class<T> dataType, String code, String message) {
        getListObservable(dataType).setValue(Resources.<List<T>>failed(code, message));
    }

    /**
     * Post one event by EventBus
     *
     * @param event the event to post
     */
    protected void post(Object event) {
        Bus.get().post(event);
    }

    /**
     * Post one sticky event by EventBus
     *
     * @param event the sticky event
     */
    protected void postSticky(Object event) {
        Bus.get().postSticky(event);
    }

    @Override
    protected void onCleared() {
        if (useEventBus) {
            Bus.get().unregister(this);
        }
    }
}

File: vmlib/src/main/java/me/shouheng/vmlib/base/BaseViewModel.kt
package me.shouheng.vmlib.base

import android.app.Application
import androidx.annotation.RestrictTo
import androidx.lifecycle.AndroidViewModel
import me.shouheng.vmlib.anno.ViewModelConfiguration
import me.shouheng.vmlib.bean.Resources
import me.shouheng.vmlib.bus.Bus
import me.shouheng.vmlib.component.IntermediateLiveData
import me.shouheng.vmlib.component.LiveDataPool
import me.shouheng.vmlib.component.ResourceLiveData

/**
 * Basic implementation of common ViewModel.
 *
 * The view model has no pre-defined model associated, for MVVMs don't want to take care of the
 * model logic. You can get the data source anywhere, which is perhaps useful for small project.
 *
 * @author [ShouhengWang](mailto:shouheng2020@gmail.com)
 * @version 2019-6-29
 */
open class BaseViewModel(application: Application) : AndroidViewModel(application) {

    private val liveDataPool = LiveDataPool<Any>()

    private val listLiveDataPool: LiveDataPool<*> = LiveDataPool<List<*>>()

    private val intermediateLiveData = IntermediateLiveData()

    /** If the view model use event bus */
    private var useEventBus = false

    /** Get a livedata of given type and sid which is used to distinguish two livedata of same type. */
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun <T> getObservable(dataType: Class<T>, sid: Int? = null): ResourceLiveData<T> {
        return liveDataPool.acquireLiveData(dataType, sid, intermediateLiveData) as ResourceLiveData<T>
    }

    /** Get a livedata of given list type and sid which is used to distinguish two livedata of same type. */
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun <T> getListObservable(dataType: Class<T>, sid: Int? = null): ResourceLiveData<List<T>> {
        return listLiveDataPool.acquireLiveData(dataType, sid, intermediateLiveData) as ResourceLiveData<List<T>>
    }

    /** Get a value base on [dataType] and [sid]. */
    fun <T> getValue(dataType: Class<T>, sid: Int? = null): T? = getObservable(dataType, sid).value?.data

    /** Get a list value base on [dataType] and [sid]. */
    fun <T> getListValue(dataType: Class<T>, sid: Int? = null): List<T>? = getListObservable(dataType, sid).value?.data

    /** Use resource to notify the value. */
    protected fun <T> setResources(
        dataType: Class<T>,
        resources: Resources<T>,
        sid: Int? = null,
    ) {
        getObservable(dataType, sid).value = resources
    }

    /** Set success state of data type */
    protected fun <T> setSuccess(
        dataType: Class<T>,
        data: T?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getObservable(dataType, sid).value = Resources.success(data, udf1, udf2, udf3, udf4, udf5) }

    /** Set progress state of data type */
    protected fun <T> setProgress(
        dataType: Class<T>,
        data: T?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getObservable(dataType, sid).value = Resources.progress(data, udf1, udf2, udf3, udf4, udf5) }

    /** Set loading state of data type */
    protected fun <T> setLoading(
        dataType: Class<T>,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getObservable(dataType, sid).value = Resources.loading(udf1, udf2, udf3, udf4, udf5) }

    /** Set fail state of data type */
    protected fun <T> setFailure(
        dataType: Class<T>,
        code: String?,
        message: String?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getObservable(dataType, sid).value = Resources.failure(code, message, udf1, udf2, udf3, udf4, udf5) }

    /** Use resource to notify the value. */
    protected fun <T> setListResources(
        dataType: Class<T>,
        resources: Resources<List<T>>,
        sid: Int?       = null,
    ) {
        getListObservable(dataType, sid).value = resources
    }

    /** Set success state of list data type */
    protected fun <T> setListSuccess(
        dataType: Class<T>,
        data: List<T>?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getListObservable(dataType, sid).value = Resources.success(data, udf1, udf2, udf3, udf4, udf5) }

    /** Set progress state of list data type */
    protected fun <T> setListProgress(
        dataType: Class<T>,
        data: List<T>?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getListObservable(dataType, sid).value = Resources.progress(data, udf1, udf2, udf3, udf4, udf5) }

    /** Set loading state of list data type */
    protected fun <T> setListLoading(
        dataType: Class<T>,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getListObservable(dataType, sid).value = Resources.loading(udf1, udf2, udf3, udf4, udf5) }

    /** Set fail state of list data type */
    protected fun <T> setListFailure(
        dataType: Class<T>,
        code: String?,
        message: String?,
        sid: Int?       = null,
        udf1: Long?     = null,
        udf2: Double?   = null,
        udf3: Boolean?  = null,
        udf4: String?   = null,
        udf5: Any?      = null
    ) { getListObservable(dataType, sid).value = Resources.failure(code, message, udf1, udf2, udf3, udf4, udf5) }

    override fun onCleared() {
        if (useEventBus) {
            Bus.get().unregister(this)
        }
    }

    init {
        val configuration = this.javaClass.getAnnotation(ViewModelConfiguration::class.java)
        if (configuration != null) {
            useEventBus = configuration.useEventBus
            if (useEventBus) {
                Bus.get().register(this)
            }
        }
    }
}
